src/Neo.VM/ReferenceCounter.cs

```diff
@@ -146,10 +148,16 @@ internal void RemoveReference(StackItem item, CompoundType parent)
 
         internal void RemoveStackReference(StackItem item)
         {
-            references_count--;
+            DecrementReferences();
             if (!NeedTrack(item)) return;
             if (--item.StackReferences == 0)
                 zero_referred.Add(item);
         }
+
+        private void DecrementReferences(int count = 1)
+        {
+            references_count -= count;
+            if (references_count < 0) throw new InvalidOperationException("Reference count is negative.");
+        }
```

Comment:
This change is controversial. As I've said in https://github.com/neo-project/neo/pull/3301#discussion_r1631061668, normally reference counter is not expected to be negative. And if it's negative then it's a bug somewhere in handlers implementation (like some stackitem not being copied properly in https://github.com/neo-project/neo/pull/3301 or something like that) which leads to the negative reference counter.

We should fix the root of the problem, not the consequence of it. To me, it should be a denial of service in case of negative reference counter, not the FAULTed transaction. Because with the current version no one will ever know about the fact that node faced with the negative reference counter issue until we go through *all* FAULTed transactions and check their FAULT exceptions.

If we replace the current fix with DoS, then the node operator will know about the fact that the node does not count references properly and coredevs will be aware of this problem and need to find the exact place in the handlers code that leads to the negative refcount.

And finally, I'd say that the most suitable case for the *uncatchable* exception is in https://github.com/neo-project/neo/blob/38cc0e9384a31f3b492225a190c38aba72d7c7f9/src/Neo.VM/ExecutionEngine.cs#L282

Reply:
I understand what you mean clearly, but what i want to know is why cant we fix this problem?

ReferenceCounter is a standalone and complete class, non-negative should be its critical property from the very first place, why cant i check it?

> We should fix the root of the problem, not the consequence of it. 

One module should not rely on the correct usage to ensure its correct execution. So from my view, both reference counter and the deepcopy are the root of the problem. consequence is DOS attack.

> If we replace the current fix with DoS, then the node operator will know about the fact that the node does not count references properly and coredevs will be aware of this problem and need to find the exact place in the handlers code that leads to the negative refcount.

To be honest, i dont quite understand. @shargon how do you think?

Reply:
> why cant we fix this problem?

We can fix it, but not in this way.

> why cant i check it?

We can check it. But what I suggest is to throw an uncatchable exception in case of a negative reference counter (the exception that won't be catched by VM). Because this exception is a direct consequence of a bug in the implementation. And if one day this exception is triggered, then we need to go and find the root of it and fix it instead of hiding this problem under catchable exception that will FAULT the transaction and allow the node to continue blocks processing.

Reply:
> I prefer an exception instead of a denial of service

It's all about it, because in this case I prefer denial of service instead of FAULTed transaction, to be able to fix the real bug in the implementation if this problem with negative refcount happens again due to some other reason. Because it's not about VM limits, it's about bugs in the implementation.

@shargon, what do you think?

Reply:
what problem of this fix can cause? you think its not necessary or you think it is not necessaey? is negative counter being designed this way?

Reply:
> what problem of this fix can cause?

This fix can cause the following problem: if there are other bugs in the node implementation that make VM counter negative, then we likely will never know about them. This fix hides the original problem, it makes the transaction FAULTed and almost no one from core devs tracks FAULTed transactions. How would we know that refcount may become negative in this case?

> you think its not necessary or you think it should not being fixed?

Neither of it. I think it should be replaced with another solution.

> is negative counter being designed this way?

No, and that's exactly the reason I suggest the denial of service instead of catchable exception that FAULTs transaction.

Reply:
Jimmy, I see that we just have different preferences choosing between DoS and FAULTed transaction. Can we have some other opinions from the @neo-project/core? If other members agree with the current implementation, then probably my PoW isn't correct.

One thing that I'd like to highlight here about the current implementation: we need to check that at least in mainnet this fix does not cause changes in the node state. It may be possible if there is some other transaction in the mainnet history that makes refcount negative (due to some other reason than in https://github.com/neo-project/neo/pull/3301).